Source/CarthageKit/DiagnosticResolver.swift

```diff
@@ -0,0 +1,170 @@
+
+	// Specify mappings to anonimize private dependencies (which may not be disclosed as part of the diagnostics)
+	public var dependencyMappings: [Dependency: Dependency]?
+
+	public init(
+		versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>,
+		dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>,
+		resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+		) {
+		self.versionsForDependency = versionsForDependency
+		self.dependenciesForDependency = dependenciesForDependency
+		self.resolvedGitReference = resolvedGitReference
+	}
+
+	public func resolve(
+		dependencies: [Dependency: VersionSpecifier],
+		lastResolved: [Dependency: PinnedVersion]? = nil,
+		dependenciesToUpdate: [String]? = nil
+		) -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
+		let result: Result<[Dependency: PinnedVersion], CarthageError>
+
+		do {
+			try traverse(dependencies: Array(dependencies))
+			result = .success([Dependency: PinnedVersion]())
+		} catch let error {
+			let carthageError: CarthageError = (error as? CarthageError) ?? CarthageError.internalError(description: error.localizedDescription)
+			result = .failure(carthageError)
+		}
+		return SignalProducer(result: result)
+	}
+
+	private func traverse(dependencies: [(Dependency, VersionSpecifier)]) throws {
+		for (dependency, versionSpecifier) in dependencies {
+			let versionSet = try findAllVersions(for: dependency, compatibleWith: versionSpecifier)
+			for version in versionSet {
+				let pinnedDependency = PinnedDependency(dependency: dependency, pinnedVersion: version)
+
+				if !handledDependencies.contains(pinnedDependency) {
+					handledDependencies.insert(pinnedDependency)
+
+					let transitiveDependencies = try findDependencies(for: dependency, version: version)
+					try traverse(dependencies: transitiveDependencies)
+				}
+			}
+		}
+	}
+
+	private func findAllVersions(for dependency: Dependency, compatibleWith versionSpecifier: VersionSpecifier) throws -> [PinnedVersion] {
+		do {
+			let versionSet: [PinnedVersion]
+			if let cachedVersionSet = versionSets[dependency] {
+				versionSet = cachedVersionSet
+			} else {
+				var pinnedVersions = [PinnedVersion]()
+				let pinnedVersionsProducer: SignalProducer<PinnedVersion, CarthageError>
+				var gitReference: String?
+
+				switch versionSpecifier {
+				case .gitReference(let hash):
+					pinnedVersionsProducer = resolvedGitReference(dependency, hash)
+					gitReference = hash
+				default:
+					pinnedVersionsProducer = versionsForDependency(dependency)
+				}
+
+				let concreteVersionsProducer = pinnedVersionsProducer.filterMap { pinnedVersion -> PinnedVersion? in
+					pinnedVersions.append(pinnedVersion)
+					return nil
+				}
+
+				_ = try concreteVersionsProducer.collect().first()!.dematerialize()
```

Comment:
Hmm? I have not tested yet but this should crash as `filterMap` always returns `nil` so far..